File: 351002/Stepanenko/lab6/src/main/java/by/bsuir/distcomp/core/security/JwtTokenProvider.java
package by.bsuir.distcomp.core.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;

@Component
public class JwtTokenProvider {

    @Value("${jwt.secret}")
    private String secret;

    @Value("${jwt.expiration}")
    private long validityInMilliseconds;

    private SecretKey key;

    @PostConstruct
    protected void init() {
        key = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
    }

    public String createToken(String login, String role) {
        Date now = new Date();
        Date validity = new Date(now.getTime() + validityInMilliseconds);

        return Jwts.builder()
                .subject(login)
                .claim("role", role)
                .issuedAt(now)
                .expiration(validity)
                .signWith(key)
                .compact();
    }

    public String getLogin(String token) {
        return getClaims(token).getSubject();
    }

    public String getRole(String token) {
        return getClaims(token).get("role", String.class);
    }

    public boolean validateToken(String token) {
        try {
            return getClaims(token).getExpiration().after(new Date());
        } catch (Exception e) {
            return false;
        }
    }

    private Claims getClaims(String token) {
        return Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/security/JwtTokenProvider.kt
package com.distcomp.security

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Component
import java.util.Date
import javax.crypto.SecretKey

@Component
class JwtTokenProvider {
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(
        "very-secret-key-at-least-32-chars!!".toByteArray()
    )
    private val expirationMs = 86_400_000L

    fun generateToken(login: String, role: String): String {
        val now = Date()
        return Jwts.builder()
            .subject(login)
            .claim("role", role)
            .issuedAt(now)
            .expiration(Date(now.time + expirationMs))
            .signWith(secretKey)
            .compact()
    }

    fun getLogin(token: String): String =
        getClaims(token).subject

    fun getRole(token: String): String =
        getClaims(token).get("role", String::class.java)

    fun isValid(token: String): Boolean = runCatching {
        getClaims(token)
        true
    }.getOrDefault(false)

    private fun getClaims(token: String) =
        Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).payload
}